### DISCUSSION TITLE
Breaking changes in Rsbuild v1.0.0

### DISCUSSION THREAD

**Discussion by chenjiahan** [web-infra-dev/rsbuild#2508] [RFC Discussions]
Rsbuild v1.0.0 is scheduled to be released in July 2024.

This thread is used to collect planned breaking changes for Rsbuild v1.0.0.

**Comment by chenjiahan** [comment]
## Upgrade Rspack v1.0.0

Rsbuild v1.0.0 will upgrade the dependent `@rspack/core` to v1.0.0.

Rspack v1.0.0 contains some breaking changes, see https://github.com/web-infra-dev/rspack/discussions/6626 for more details.

**Comment by chenjiahan** [comment]
## Rename `dev.startUrl`

Changes:

- Rename `dev.startUrl` to `server.open`.
- Rename `dev.beforeStartUrl` to `server.open.before`.

Motivation:

- Align with Rspack dev server's [open](https://www.rspack.dev/config/dev-server#devserveropen) config.
- Allow to use in the preview server.

**Reply by chenjiahan** [reply]
https://github.com/web-infra-dev/rsbuild/pull/2642

**Comment by 9aoy** [comment]
## Replace `output.targets` with `output.target`

Replace `output.targets` with `output.target`, and rsbuild will generate rspack / webpack configs based on environments instead of `output.targets`.

Details: https://github.com/web-infra-dev/rsbuild/pull/2636

**Comment by chenjiahan** [comment]
## output.overrideBrowserslist

`output.overrideBrowserslist` no longer supports `Record<RsbuildTarget, string[]` type for environment configuration.

- before:

```js
export default {
  output: {
    overrideBrowserslist: {
      web: ['iOS >= 9', 'Android >= 4.4'],
      node: ['node >= 20'],
    },
  },
};
```

- after:

```js
export default defineConfig({
  environments: {
    web: {
      output: {
        overrideBrowserslist: ['iOS >= 9', 'Android >= 4.4'],
      },
    },
    node: {
      output: {
        overrideBrowserslist: ['node >= 20'],
      },
    },
  },
});
```

**Reply by chenjiahan** [reply]
https://github.com/web-infra-dev/rsbuild/pull/2646

**Reply by chenjiahan** [reply]
https://github.com/web-infra-dev/rsbuild/pull/2648

**Comment by chenjiahan** [comment]
## output.emitAssets

`output.emitAssets` changed to boolean type for environment configuration.

- before:

```js
export default {
  output: {
    targets: ['web', 'node'],
    emitAssets: ({ target }) => target !== 'node',
  },
};
```

- after:

```js
export default {
  environments: {
    web: {
      output: {
        target: 'web',
      },
    },
    node: {
      output: {
        target: 'node',
        emitAssets: false,
      },
    },
  },
};
```

**Reply by chenjiahan** [reply]
https://github.com/web-infra-dev/rsbuild/pull/2649

**Comment by chenjiahan** [comment]
## source.alias

Remove `target` param for `source.alias` function, use environment configuration instead.

- before:

```js
export default {
  source: {
    alias: (alias, { target }) => {
      if (target === 'node') {
        alias['@common'] = './src/node/common';
      } else if (target === 'web') {
        alias['@common'] = './src/web/common';
      }
    },
  },
  output: {
    targets: ['web', 'node'],
  }
};
```

- after:

```js
export default {
  environments: {
    web: {
      source: {
        alias: {
          '@common': './src/web/common',
        },
      },
      output: {
        target: 'web',
      },
    },
    node: {
      source: {
        alias: {
          '@common': './src/node/common',
        },
      },
      output: {
        target: 'node',
      },
    },
  },
};
```

**Reply by chenjiahan** [reply]
https://github.com/web-infra-dev/rsbuild/pull/2650

**Comment by chenjiahan** [comment]
## Decorators Version

Rsbuild now uses `2022-11` decorators version by default. This allows Rsbuild to align the default behavior with TypeScript >= 5.0 and esbuild >= 0.21.0.

If you are using the legacy decorators, you can add the following config:

```ts
// rsbuild.config.ts
import { defineConfig } from '@rsbuild/core';

export default defineConfig({
  source: {
    decorators: {
      version: 'legacy',
    },
  },
});
```

SWC currently does not support `2023-11`, when this PR (https://github.com/swc-project/swc/pull/8970) is merged, we can consider use `2023-11` by default.

**Reply by Verthon** [reply]
SWC closed linked PR https://github.com/swc-project/swc/pull/8970 in favor of https://github.com/swc-project/swc/pull/9488

**Comment by chenjiahan** [comment]
## distPath.server

Removed `output.distPath.server`, use the `environments` config instead.

```js
export default {
  environments: {
    web: {
      output: {
        target: 'web',
      },
    },
    node: {
      output: {
        target: 'node',
        distPath: {
          root: 'dist/server',
        },
      },
    },
  },
};
```

**Reply by chenjiahan** [reply]
https://github.com/web-infra-dev/rsbuild/pull/2696

**Comment by chenjiahan** [comment]
## source.entry

Remove function usage of `source.entry`, use `environments` config instead.

- before:

```js
export default {
  source: {
    entry({ target }) {
      if (target === 'web') {
        return {
          index: './src/index.client.js',
        };
      }
      if (target === 'node') {
        return {
          index: './src/index.server.js',
        };
      }
    },
  },
  output: {
    targets: ['web', 'node'],
  },
};
```

- after:

```js
export default {
  environments: {
    web: {
      source: {
        entry: {
          index: './src/index.client.js',
        },
      },
      output: {
        target: 'web',
      },
    },
    node: {
      source: {
        entry: {
          index: './src/index.server.js',
        },
      },
      output: {
        target: 'node',
      },
    },
  },
};
```

**Reply by chenjiahan** [reply]
https://github.com/web-infra-dev/rsbuild/pull/2697

**Comment by 9aoy** [comment]
## api.getHTMLPaths

Remove `api.getHtmlPath`, use environment htmlPaths  instead.

- before:

```ts
const pluginFoo = () => ({
  setup(api) {
    api.modifyRspackConfig(() => {
      const htmlPaths = api.getHTMLPaths();
      console.log(htmlPaths); // { index: 'index.html' };
    });
  },
});
```

- after:
```ts
const pluginFoo = () => ({
  setup(api) {
    api.modifyRspackConfig((config, { environment }) => {
      const { htmlPaths } = environment;
      console.log(htmlPaths); // { index: 'index.html' };
    });
  },
});
```

**Reply by 9aoy** [reply]
https://github.com/web-infra-dev/rsbuild/pull/2771

**Comment by 9aoy** [comment]
## api.context
Remove `api.context.entry` and `api.context.targets`, and the `distPath` and `tsconfigPath` may be different in different environments, should use environment / environments param instead.

```diff
const pluginFoo = () => ({
  setup(api) {
      api.modifyBundlerChain((chain, { environment }) => {
-        const { entry,  targets, tsconfigPath, distPath } = api.context;
+        const { entry, tsconfigPath, distPath, config: { output: { target } } } = environment;
    });
    api.onDevCompileDone(async ({ isFirstCompile, stats, environments }) => {
-      const entries = Object.keys(api.context.entry);
+      const entries = Object.values(environments).map(e => e.entry);
    })
  },
});
```

**Reply by 9aoy** [reply]
https://github.com/web-infra-dev/rsbuild/pull/2680  
https://github.com/web-infra-dev/rsbuild/pull/2721 
https://github.com/web-infra-dev/rsbuild/pull/2768

**Comment by chenjiahan** [comment]
## output.polyfill

Rsbuild 1.0.0-alpha.4 set `output.polyfill` to `'off'` by default, this can reduce the polyfill code and generate smaller bundles by default.

If your application need polyfill, please set `output.polyfill` to `'usage'` or `'entry'`:

```ts
// rsbuild.config.ts
export default {
  output: {
    polyfill: 'usage'
  }
};
```

See the vote on Twitter: https://x.com/rspack_dev/status/1808439138439274780

![image](https://github.com/web-infra-dev/rsbuild/assets/7237365/f316e9e6-94e2-4c3d-8216-3d62626fcd02)

**Reply by chenjiahan** [reply]
https://github.com/web-infra-dev/rsbuild/pull/2804

**Comment by chenjiahan** [comment]
## Remove `output.minify.html`

Rsbuild `1.0.0-alpha.6` removed the `output.minify.html` and `output.minify.htmlOptions` options, and no longer minify the HTML files.

Previously Rsbuild uses `html-minifier-terser` to minify the HTML files. But the performance of `html-minifier-terser` can not meet the needs of Rsbuild applications, and In most cases, there is little benefit in compressing HTML. 

At this stage, we do not want to builtin `html-minifier-terser` , so we provide a standalone [rsbuild-plugin-html-minifier-terser](https://github.com/rspack-contrib/rsbuild-plugin-html-minifier-terser) to support HTML minification.

```js
// rsbuild.config.ts
import { pluginHtmlMinifierTerser } from "rsbuild-plugin-html-minifier-terser";

export default {
  plugins: [pluginHtmlMinifierTerser()],
};
```

And we plan to use some faster Rust-based HTML minimizer in the future.

**Reply by chenjiahan** [reply]
https://github.com/web-infra-dev/rsbuild/pull/2833

**Comment by chenjiahan** [comment]
## output.charset

The default value of [output.charset](https://rsbuild.dev/config/output/charset) has been changed from ascii to utf8.

If you need to use ascii, you can add the following config:

```js
export default {
  output: {
    charset: 'ascii',
  },
};
```

**Reply by chenjiahan** [reply]
See https://github.com/web-infra-dev/rsbuild/discussions/3244 for more details
